### PR TITLE
Fix a bunch of RS1024 on Equals

### PR DESCRIPTION
**Pull request by Evangelink**
Not sure why but I have started to get hundreds of warning related to `RS1024` in the codebase so I decided to start tackling them.

In this PR, I am focusing on the `Equals` because for some reason most of the complains on collections don't get fixed after applying the comparer.

### CODE REVIEW

File: src/Utilities/Compiler/SymbolEqualityComparer.cs [hunk 1]
```diff
@@ -0,0 +1,21 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+namespace Microsoft.CodeAnalysis
+{
+#if !CODEANALYSIS_V3_OR_BETTER
```
Comment (Evangelink) [inline src/Utilities/Compiler/SymbolEqualityComparer.cs:5]: I had some exceptions saying cannot find `SymbolEqualityComparer` which made me realize that it's available only since v3. Adding a class matching the symbol to fallback to object equality seems the simplest solution as it will prevent pragmas everywhere in the code and make it simple when we drop support for older version to "bump" to the new version.

I am not entirely sure why but this symbol is not giving me what I would except so I end up with some warnings telling me that the symbol exists in 2 locations.

Any idea which compilation symbol to use to have the "proper" fix?

File: src/Utilities/Compiler/WellKnownTypeProvider.cs [hunk 2]
```diff
@@ -30,6 +30,9 @@ private WellKnownTypeProvider(Compilation compilation)
                 () =>
                 {
                     return ImmutableHashSet.Create<IAssemblySymbol>(
+#if CODEANALYSIS_V3_OR_BETTER
```
Comment (Evangelink) [inline src/Utilities/Compiler/WellKnownTypeProvider.cs:33]: Need to revert this change to use the added class.

File: src/Utilities/Compiler/WellKnownTypeProvider.cs [hunk 3]
```diff
@@ -200,7 +203,11 @@ internal bool IsTaskOfType([NotNullWhen(returnValue: true)] ITypeSymbol? typeSym
         {
             return typeSymbol != null
                 && typeSymbol.OriginalDefinition != null
-                && typeSymbol.OriginalDefinition.Equals(GetOrCreateTypeByMetadataName(WellKnownTypeNames.SystemThreadingTasksTask1))
+                &&
+#if CODEANALYSIS_V3_OR_BETTER
```
Comment (Evangelink) [inline src/Utilities/Compiler/WellKnownTypeProvider.cs:207]: Need to revert this change to use the added class.

File: src/Utilities/Compiler/Extensions/IMethodSymbolExtensions.cs [hunk 4]
```diff
@@ -496,7 +500,7 @@ public static bool IsCollectionAddMethod(this IMethodSymbol method, ImmutableHas
         /// <param name="taskType">Task type.</param>
         public static bool IsTaskFromResultMethod(this IMethodSymbol method, [NotNullWhen(returnValue: true)] INamedTypeSymbol? taskType)
             => method.Name.Equals("FromResult", StringComparison.Ordinal) &&
-               method.ContainingType.Equals(taskType);
+               SymbolEqualityComparer.Default.Equals(method.ContainingType, taskType);
```
Comment (Evangelink) [inline src/Utilities/Compiler/Extensions/IMethodSymbolExtensions.cs:503]: Shall we check the `OriginalDefinition` here?

File: src/Microsoft.CodeAnalysis.Analyzers/CSharp/MetaAnalyzers/CSharpRegisterActionAnalyzer.cs [hunk 5]
```diff
@@ -78,8 +78,8 @@ protected override SyntaxNode GetInvocationExpression(InvocationExpressionSyntax
 
             protected override bool IsSyntaxKind(ITypeSymbol type)
             {
-                return (_csharpSyntaxKind != null && type.Equals(_csharpSyntaxKind)) ||
-                    (_basicSyntaxKind != null && type.Equals(_basicSyntaxKind));
+                return (_csharpSyntaxKind != null && SymbolEqualityComparer.Default.Equals(type, _csharpSyntaxKind)) ||
+                    (_basicSyntaxKind != null && SymbolEqualityComparer.Default.Equals(type, _basicSyntaxKind));
```
Comment (Youssef1313) [inline src/Microsoft.CodeAnalysis.Analyzers/CSharp/MetaAnalyzers/CSharpRegisterActionAnalyzer.cs:82]: I believe the null check is redundant

File: src/Microsoft.CodeAnalysis.Analyzers/Core/FixAnalyzers/FixerWithFixAllAnalyzer.cs [hunk 6]
```diff
@@ -248,7 +248,7 @@ internal void CompilationEnd(CompilationAnalysisContext context)
                     {
                         AnalyzeFixerWithFixAll(fixer, context);
                     }
-                    else if (fixer.BaseType != null && fixer.BaseType.Equals(_codeFixProviderSymbol))
+                    else if (fixer.BaseType != null && SymbolEqualityComparer.Default.Equals(fixer.BaseType, _codeFixProviderSymbol))
```
Comment (Youssef1313) [inline src/Microsoft.CodeAnalysis.Analyzers/Core/FixAnalyzers/FixerWithFixAllAnalyzer.cs:251]: The null check was previously needed, but now it's redundant.

File: src/Microsoft.CodeAnalysis.Analyzers/Core/FixAnalyzers/FixerWithFixAllAnalyzer.cs [hunk 7]
```diff
@@ -335,13 +335,13 @@ bool IsViolatingCodeActionObjectCreation(IObjectCreationOperation objectCreation
                     // Local functions
                     bool IsCodeActionWithOverriddenEquivalenceKey(INamedTypeSymbol namedType)
                     {
-                        if (namedType == null || namedType.Equals(_codeActionSymbol))
+                        if (namedType == null || SymbolEqualityComparer.Default.Equals(namedType, _codeActionSymbol))
```
Comment (Youssef1313) [inline src/Microsoft.CodeAnalysis.Analyzers/Core/FixAnalyzers/FixerWithFixAllAnalyzer.cs:338]: Null check is now redundant

File: src/Microsoft.CodeAnalysis.Analyzers/Core/MetaAnalyzers/CompareSymbolsCorrectlyAnalyzer.cs [hunk 8]
```diff
@@ -269,7 +269,7 @@ private static bool IsSymbolType(ITypeSymbol typeSymbol, INamedTypeSymbol symbol
                 return false;
             }
 
-            if (typeSymbol.Equals(symbolType))
+            if (SymbolEqualityComparer.Default.Equals(typeSymbol, symbolType))
```
Comment (Youssef1313) [inline src/Microsoft.CodeAnalysis.Analyzers/Core/MetaAnalyzers/CompareSymbolsCorrectlyAnalyzer.cs:272]: null check above is now redundant.

File: src/Utilities/Compiler/Extensions/IMethodSymbolExtensions.cs [hunk 9]
```diff
@@ -213,7 +214,9 @@ public static bool IsAsyncDisposeImplementation([NotNullWhen(returnValue: true)]
 
             // Identify the implementor of IAsyncDisposable.Dispose in the given method's containing type and check
             // if it is the given method.
-            return method.ReturnType != null && method.ReturnType.Equals(valueTaskType) && method.Parameters.IsEmpty &&
+            return method.ReturnType != null &&
```
Comment (Youssef1313) [inline src/Utilities/Compiler/Extensions/IMethodSymbolExtensions.cs:217]: null check is now redundant.

File: src/Utilities/Compiler/Extensions/IMethodSymbolExtensions.cs (hunk 4 above)
Comment (Youssef1313) [reply src/Utilities/Compiler/Extensions/IMethodSymbolExtensions.cs:503]: > Shall we check the `OriginalDefinition` here?

This is the non-generic task type. So I think the way it's currently written is okay.

File: src/Utilities/Compiler/Extensions/ISymbolExtensions.cs [hunk 10]
```diff
@@ -713,7 +713,7 @@ public static IEnumerable<AttributeData> GetAttributes(this ISymbol symbol, INam
                 return Enumerable.Empty<AttributeData>();
             }
 
-            return symbol.GetAttributes().Where(attr => attr.AttributeClass.Equals(attributeType));
+            return symbol.GetAttributes().Where(attr => SymbolEqualityComparer.Default.Equals(attr.AttributeClass, attributeType));
```
Comment (Youssef1313) [inline src/Utilities/Compiler/Extensions/ISymbolExtensions.cs:716]: You can use HasAttribute extension method here instead.

File: src/Utilities/Compiler/Extensions/OperationBlockAnalysisContextExtension.cs [hunk 11]
```diff
@@ -53,10 +53,16 @@ static bool IsSingleStatementBody(IBlockOperation body)
                     descendants[0] is IThrowOperation throwOperation &&
                     throwOperation.GetThrownExceptionType() is ITypeSymbol createdExceptionType)
                 {
-                    if (Equals(context.Compilation.GetOrCreateTypeByMetadataName(WellKnownTypeNames.SystemNotImplementedException), createdExceptionType.OriginalDefinition) ||
-                        Equals(context.Compilation.GetOrCreateTypeByMetadataName(WellKnownTypeNames.SystemNotSupportedException), createdExceptionType.OriginalDefinition) ||
+                    if (SymbolEqualityComparer.Default.Equals(
+                            context.Compilation.GetOrCreateTypeByMetadataName(WellKnownTypeNames.SystemNotImplementedException),
```
Comment (Youssef1313) [inline src/Utilities/Compiler/Extensions/OperationBlockAnalysisContextExtension.cs:57]: Out of scope for this PR, but I'm worried about the performance when `GetOrCreateTypeByMetadataName` is called multiple times in this extension.

I think it may be better to get the symbols once at compilation start for the analyzers using this extension, and pass them to the method.

File: src/Utilities/Compiler/WellKnownTypeProvider.cs [hunk 12]
```diff
@@ -200,7 +203,11 @@ internal bool IsTaskOfType([NotNullWhen(returnValue: true)] ITypeSymbol? typeSym
...
+                    SymbolEqualityComparer.Default.
+#endif
+                    Equals(typeSymbol.OriginalDefinition, GetOrCreateTypeByMetadataName(WellKnownTypeNames.SystemThreadingTasksTask1))
```
Comment (Youssef1313) [inline src/Utilities/Compiler/WellKnownTypeProvider.cs:210]: Same concern about performance if the extension is called multiple times in the same analyzer.

File: src/Utilities/Compiler/SymbolEqualityComparer.cs (hunk 1 above)
Comment (Youssef1313) [reply src/Utilities/Compiler/SymbolEqualityComparer.cs:5]: > I am not entirely sure why but this symbol is not giving me what I would except so I end up with some warnings telling me that the symbol exists in 2 locations.


@Evangelink I'm not seeing anything about that in CI.

File: src/Utilities/Compiler/Extensions/IMethodSymbolExtensions.cs [hunk 13]
```diff
@@ -18,7 +18,7 @@
 
 namespace Analyzer.Utilities.Extensions
 {
-    internal static class IMethodSymbolExtensions
+    internal static partial class IMethodSymbolExtensions
```
Comment (Youssef1313) [inline src/Utilities/Compiler/Extensions/IMethodSymbolExtensions.cs:21]: Why this change?

File: src/Utilities/Compiler/SymbolEqualityComparer.cs [hunk 14]
```diff
@@ -0,0 +1,21 @@
...
+    internal class SymbolEqualityComparer
```
Comment (Youssef1313) [inline src/Utilities/Compiler/SymbolEqualityComparer.cs:6]:
```suggestion
    internal sealed class SymbolEqualityComparer : IEqualityComparer<ISymbol?>
```

File: src/Utilities/Compiler/SymbolEqualityComparer.cs [hunk 15]
```diff
@@ -0,0 +1,21 @@
...
+    {
+        private SymbolEqualityComparer()
+        {
+
+        }
+
+#pragma warning disable CA1822 // Mark members as static
+        public bool Equals(ISymbol? symbol1, ISymbol? symbol2) =>
+#pragma warning restore CA1822 // Mark members as static
+            object.Equals(symbol1, symbol2);
```
Comment (Youssef1313) [inline src/Utilities/Compiler/SymbolEqualityComparer.cs:16]: To be more matching with roslyn implementation

```suggestion
        public bool Equals(ISymbol? x, ISymbol? y)
        {
            if (x is null)
            {
                return y is null;
            }

            return x.Equals(y);
        }

        public int GetHashCode(ISymbol? obj)
        {
            return obj?.GetHashCode() ?? 0;
        }
```

Comment (Evangelink) [reply src/Utilities/Compiler/SymbolEqualityComparer.cs:16]: Thanks!

File: src/Utilities/Compiler/Extensions/OperationBlockAnalysisContextExtension.cs (hunk 11 above)
Comment (Evangelink) [reply src/Utilities/Compiler/Extensions/OperationBlockAnalysisContextExtension.cs:57]: I think that perf should be ok because we are doing a dictionary call (if it exists we return the value otherwise we create, store and return). We implemented this helper especially to help with performance issue.

File: src/Utilities/Compiler/SymbolEqualityComparer.cs (hunk 1 above)
Comment (Evangelink) [reply src/Utilities/Compiler/SymbolEqualityComparer.cs:5]: I can no longer reproduce the issue. I am wondering if it was just some kind of non-refresh from VS.

**Comment by Evangelink** [comment]
Ping @mavasani

File: src/Utilities/Compiler/Extensions/ISymbolExtensions.cs (hunk 10 above)
Comment (Evangelink) [reply src/Utilities/Compiler/Extensions/ISymbolExtensions.cs:716]: Just had a look and I can't make this replacement here.

File: src/Utilities/Compiler/Extensions/OperationBlockAnalysisContextExtension.cs (hunk 11 above)
Comment (mavasani) [reply src/Utilities/Compiler/Extensions/OperationBlockAnalysisContextExtension.cs:57]: Both of you are right to an extent - perf is much better even inside operation callbacks because it is a dictionary, however we still prefer to collect all well known symbols once at compilation start and pass them around, primary for consistency in our code base and also avoid large number of operation callbacks trying to access the same dictionary when multiple analyzers are executing in parallel.

Comment (mavasani) [reply src/Utilities/Compiler/Extensions/OperationBlockAnalysisContextExtension.cs:57]: Agreed this is our of scope for this PR though.

File: src/Utilities/Compiler/SymbolEqualityComparer.cs [hunk 16]
```diff
@@ -0,0 +1,26 @@
...
+#if !CODEANALYSIS_V3_OR_BETTER
+namespace Microsoft.CodeAnalysis
+{
+    using System.Collections.Generic;
+
+    internal sealed class SymbolEqualityComparer : IEqualityComparer<ISymbol?>
+    {
+        private SymbolEqualityComparer()
+        {
+
```
Comment (mavasani) [inline src/Utilities/Compiler/SymbolEqualityComparer.cs:12]:
```suggestion
```

**Review by mavasani — APPROVED** [review]

File: src/Utilities/Compiler/SymbolEqualityComparer.cs (hunk 16 above)
Comment (mavasani) [reply src/Utilities/Compiler/SymbolEqualityComparer.cs:12]: @Evangelink going to commit this minor style change and merge your PR.

**Comment by Evangelink** [comment]
Thanks!
